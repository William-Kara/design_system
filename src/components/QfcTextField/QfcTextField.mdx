import { ArgTypes, Meta } from '@storybook/blocks';
import { Story } from '@storybook/addon-docs';
import QfcTextField from '@/components/QfcTextField/QfcTextField.vue';

<Meta title="Elements/Inputs/TextField" />

# Definition

A customizable text field component This component is a text field that can be used for various types of input, such as text, password, email and number.

## Props

The following props are available for the QfcTextField component:

<ArgTypes of={QfcTextField} />
When the isRequired prop is set to true, an asterisk appears to the right of the label.

An eye icon to toggle the visibility of the input when the type is set to “password” is available.

The input field become red when the errorMessage prop is set. But also when the input get the :invalid pseudo-class that can happen when the input is required and empty or when the input doesn't match the pattern asked.

# Usage

To use QfcTextField, style needs to be imported separately.

```sass
@import '@website/design-system-vuejs/dist/styles/components/QfcTextField.scss';

```

```html
<template>
  <QfcTextField label="Name" :isRequired="true" />
</template>
```

# Declinaison

## Basic

<Story name="Basic" id="elements-inputs-textfield--primary" />

## Error

<Story name="Error" id="elements-inputs-textfield--error" />

## Disabled

<Story name="Disabled" id="elements-inputs-textfield--disabled" />

## Password

<Story name="Password" id="elements-inputs-textfield--password" />

## Required

<Story name="Required" id="elements-inputs-textfield--required" />

## Background color

The background color of the input field can be customized by using hexadecimal color codes.

```html
<template>
  <QfcTextField label="Name" :isRequired="true" backgroundColor="#f5f5f5" />
</template>
```

The background color act as background for textfield and the label moving up when the input is focused.
