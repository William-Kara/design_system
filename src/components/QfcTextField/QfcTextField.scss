@mixin input-invalid {
  border: $border-danger;
  color: $danger;
  + .text-field__label,
  ~ .text-field__img {
    color: $danger;
  }
}
@mixin input-active-focus {
  border: $border-active;
  color: $active;
  + .text-field__label,
  ~ .text-field__img {
    color: $active;
  }
}
@mixin input-hover {
  border: $border-hover;
  color: $text;
  + .text-field__label,
  ~ .text-field__img {
    color: $text;
  }
}
@mixin input-disabled {
  border: $border-disabled;
  cursor: not-allowed;
  + .text-field__label,
  ~ .text-field__img {
    color: $disabled-2;
    cursor: not-allowed;
  }
}

.text-field {
  position: relative;
  color: $grey-alto;
  text-align: right;
  &__img {
    position: relative;
    bottom: 2.25rem;
    right: 1rem;
  }
  &__input {
    box-sizing: border-box;
    width: 100%;
    padding: 0.75rem;
    font-size: 1rem;
    transition: border $transition-time;
    border: $border-default;
    border-radius: $br-xxs;

    &::placeholder {
      color: transparent;
    }
    &:hover,
    &:focus {
      outline: none;
    }
    &:hover,
    &:active {
      @include input-hover;
    }
    &:valid {
      color: $text;
    }
    &:focus {
      @include input-active-focus;
    }
    &.is-error {
      @include input-invalid;
    }
    &:disabled {
      @include input-disabled;
    }
    &:invalid {
      @include input-invalid;
    }
  }

  &__label {
    cursor: text;
    position: absolute;
    top: 0.688rem;
    left: 0.66rem;
    padding: 0 0.15rem;
    transition: all $transition-time ease-in-out;
  }

  &__input:focus + .text-field__label,
  &__input:not(:placeholder-shown) + .text-field__label {
    transform: translateY(-1rem);
    font-size: $fs-xxs;
    line-height: $fs-xxs;
  }

  &__error-message {
    color: $danger;
    font-size: $fs-xxs;
    padding-left: 0.85rem;
  }

  &_img {
    position: absolute;
    right: 0.5rem;
    width: 1.7rem;
    top: 0.6rem;
    cursor: pointer;
  }
}
