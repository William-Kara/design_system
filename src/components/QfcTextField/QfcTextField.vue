<template>
  <div class="text-field">
    <input
      :id="id"
      v-model="model"
      :name="label"
      :type="formatedType"
      :placeholder="label"
      :disabled="isDisabled"
      :class="inputClasses"
      :style="backgroundStyle"
      :required="isRequired"
      :maxlength="maxLength"
      :step="step"
      :min="min"
      :max="max"
      :pattern="pattern"
    />
    <label :for="id" class="text-field__label" :style="backgroundStyle">{{ formatedLabel }}</label>
    <span
      v-if="isPassword"
      class="text-field__img material-icons-outlined"
      @click="toggleVisibility"
      >{{ eyeIcon }}</span
    >
    <p class="text-field__error-message">{{ errorMessage }}</p>
  </div>
</template>

<script lang="ts" setup>
import { computed, ref } from 'vue';
import { useVModel } from '@/utils/useVModel';

const inputType = ['text', 'password', 'email', 'number'] as const;
type InputType = (typeof inputType)[number];

type Background = `#${string}`;

interface Props {
  id: string;
  type?: InputType;
  modelValue: string;
  label: string;
  errorMessage?: string;
  maxLength?: number;
  rows?: number;
  min?: number;
  max?: number;
  step?: number;
  pattern?: string;
  isRequired?: boolean;
  isDisabled?: boolean;
  backgroundColor?: Background;
}
const props = withDefaults(defineProps<Props>(), {
  modelValue: '',
  type: 'text',
  label: 'Label',
  isRequired: false,
  isDisabled: false,
  backgroundColor: '#232323',
});

const id = computed(() => {
  return props.id ? props.id : Date.now().toString(36) + Math.random().toString(36).substring(2);
});

const isPassword = props.type === 'password';
let isVisible = ref(!isPassword);
let eyeIcon = computed(() => (isVisible.value ? 'visibility' : 'visibility_off'));
const toggleVisibility = () => {
  isVisible.value = !isVisible.value;
};

const formatedType = computed(() => {
  if (props.type === 'email') return 'email';
  if (props.type === 'number') return 'number';
  return isVisible.value ? 'text' : 'password';
});
const formatedLabel = computed(() => {
  return props.isRequired ? `${props.label}*` : props.label;
});

const emit = defineEmits(['update:modelValue']);
const model = useVModel(props, emit);

const inputClasses = computed(() => {
  return {
    'text-field__input': true,
    'is-error': props.errorMessage,
  };
});
const backgroundStyle = computed(() => {
  return { 'background-color': props.backgroundColor };
});
</script>
<style lang="scss" scoped src="./QfcTextField.scss" />
