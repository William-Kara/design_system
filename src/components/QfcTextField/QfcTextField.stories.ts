import QfcTextField from './QfcTextField.vue';
import type { Meta, StoryObj } from '@storybook/vue3';

const meta: Meta<typeof QfcTextField> = {
  component: QfcTextField,
  title: 'Elements/Inputs/TextField',
  argTypes: {
    type: {
      control: 'select',
      defaultValue: 'text',
      options: ['text', 'password', 'email', 'number'],
    },
    label: {
      type: { name: 'string', required: true },
      defaultValue: 'Label',
    },
    modelValue: {
      type: { name: 'string', required: true },
      defaultValue: '',
    },
    errorMessage: {
      type: { name: 'string', required: false },
    },
    maxLength: {
      type: { name: 'number', required: false },
    },
    min: {
      type: { name: 'number', required: false },
    },
    max: {
      type: { name: 'number', required: false },
    },
    step: {
      type: { name: 'number', required: false },
    },
    pattern: {
      type: { name: 'string', required: false },
    },
    isRequired: {
      type: { name: 'boolean', required: false },
      defaultValue: false,
    },
    isDisabled: {
      type: { name: 'boolean', required: false },
      defaultValue: false,
    },
    backgroundColor: {
      type: { name: 'string', required: false },
    },
  },
  parameters: {
    backgrounds: {
      default: 'default',
      values: [{ name: 'default', value: '#232323' }],
    },
  },
};
export default meta;

type Story = StoryObj<typeof meta>;
export const Primary: Story = {
  args: {
    modelValue: '',
    type: 'text',
    label: 'Label',
    errorMessage: '',
    maxLength: 255,
    min: 0,
    max: 255,
    step: 1,
    pattern: '',
    isDisabled: false,
    isRequired: false,
    backgroundColor: '#232323',
  },
};

export const Error: Story = {
  args: {
    ...Primary.args,
    errorMessage: 'Error',
  },
};

export const Disabled: Story = {
  args: {
    ...Primary.args,
    isDisabled: true,
  },
};

export const Password: Story = {
  args: {
    ...Primary.args,
    type: `password`,
  },
};

export const Required: Story = {
  args: {
    ...Primary.args,
    isRequired: true,
  },
};
