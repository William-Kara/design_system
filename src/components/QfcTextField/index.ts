import QfcTextField from './QfcTextField.vue';

export { QfcTextField };
export default QfcTextField;
