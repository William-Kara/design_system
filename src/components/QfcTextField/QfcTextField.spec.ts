import { it, describe, beforeEach, expect } from 'vitest';

import { shallowMount } from '@vue/test-utils';
import QfcTextField from './QfcTextField.vue';

let wrapper: any = null;

const propsValues = {
  id: 'test',
  modelValue: '',
  label: 'Email',
  type: 'email',
  errorMessage: '',
  isRequired: true,
  isDisabled: false,
  backgroundColor: '#232323',
  max: 3,
  maxLength: 4,
  min: 1,
  rows: 5,
  pattern: '',
  step: 5,
};

beforeEach(() => {
  wrapper = shallowMount(QfcTextField, {
    propsData: propsValues,
  });
});

describe('QfcTextField Component', () => {
  it('should mount properly', () => {
    expect(wrapper).toBeTruthy();
  });

  it('should have correct props', () => {
    expect(wrapper.props()).toEqual(propsValues);
  });

  it('should render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should emit update event when value is changed', () => {
    const input = wrapper.find('input');
    input.setValue('test@example.com');
    expect(wrapper.emitted('update:modelValue')[0][0]).toEqual('test@example.com');
  });

  it('should display the error when an error message is given', () => {
    wrapper = shallowMount(QfcTextField, {
      propsData: {
        ...propsValues,
        errorMessage: 'Invalid email',
      },
    });

    expect(wrapper.find('.text-field__error-message').text()).toBe('Invalid email');
  });

  it('should have label ending by an aster when isRequired is true', () => {
    expect(wrapper.props().isRequired).toBe(true);
    expect(wrapper.find('.text-field__label').text()).toBe(`${propsValues.label}*`);
  });

  it('should have label without an aster when isRequired is false', () => {
    wrapper = shallowMount(QfcTextField, {
      propsData: {
        ...propsValues,
        isRequired: false,
      },
    });

    expect(wrapper.props().isRequired).toBe(false);
    expect(wrapper.find('.text-field__label').text()).toBe(propsValues.label);
  });
});
