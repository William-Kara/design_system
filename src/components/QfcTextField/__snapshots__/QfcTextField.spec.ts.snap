// Vitest Snapshot v1

exports[`QfcTextField Component > should render correctly 1`] = `
"<div data-v-90f2fe4b=\\"\\" class=\\"text-field\\"><input data-v-90f2fe4b=\\"\\" id=\\"test\\" name=\\"Email\\" type=\\"email\\" placeholder=\\"Email\\" class=\\"text-field__input\\" style=\\"background-color: rgb(35, 35, 35);\\" required=\\"\\" maxlength=\\"4\\" step=\\"5\\" min=\\"1\\" max=\\"3\\" pattern=\\"\\"><label data-v-90f2fe4b=\\"\\" for=\\"test\\" class=\\"text-field__label\\" style=\\"background-color: rgb(35, 35, 35);\\">Email*</label>
  <!--v-if-->
  <p data-v-90f2fe4b=\\"\\" class=\\"text-field__error-message\\"></p>
</div>"
`;
